iothub_client/samples/pnp/common/pnp_device_client_helpers.c

```diff
@@ -0,0 +1,100 @@
+    {
+        LogError("Failure creating Iothub device.  Hint: Check you connection string");
+        result = MU_FAILURE;
+    }
+    // Sets verbosity level
+    else if ((iothubResult = IoTHubDeviceClient_SetOption(deviceHandle, OPTION_LOG_TRACE, &enableTracing)) != IOTHUB_CLIENT_OK)
+    {
+        LogError("Unable to set logging option, error=%d", iothubResult);
+        result = MU_FAILURE;
+    }
+    // Sets the name of ModelId for this PnP device.
+    // This *MUST* be set before the client is connected to IoTHub.  We do not automatically connect when the 
+    // handle is created, but will implicitly connect to subscribe for device method and device twin callbacks below.
+    else if ((iothubResult = IoTHubDeviceClient_SetOption(deviceHandle, OPTION_MODEL_ID, modelId)) != IOTHUB_CLIENT_OK)
+    {
+        LogError("Unable to set the ModelID, error=%d", iothubResult);
+        result = MU_FAILURE;
+    }
+    // Sets the callback function that processes incoming device methods, which are the channel PnP Commands are transferred over
+    else if ((iothubResult = IoTHubDeviceClient_SetDeviceMethodCallback(deviceHandle, deviceMethodCallback, NULL)) != IOTHUB_CLIENT_OK)
+    {
+        LogError("Unable to set device method callback, error=%d", iothubResult);
+        result = MU_FAILURE;
+    }
+    // Sets the callback function that processes device twin changes from the hub, which is the channel that PnP Properties are 
+    // transferred over.  This will also automatically retrieve the full twin for the application. 
+    else if ((iothubResult = IoTHubDeviceClient_SetDeviceTwinCallback(deviceHandle, deviceTwinCallback, (void*)deviceHandle)) != IOTHUB_CLIENT_OK)
+    {
+        LogError("Unable to set device twin callback, error=%d", iothubResult);
+        result = MU_FAILURE;
+    }
+    // Enabling auto url encode will have the underlying SDK perform URL encoding operations automatically.
+    else if ((iothubResult = IoTHubDeviceClient_SetOption(deviceHandle, OPTION_AUTO_URL_ENCODE_DECODE, &urlAutoEncodeDecode)) != IOTHUB_CLIENT_OK)
+    {
+        LogError("Unable to set auto Url encode option, error=%d", iothubResult);
+        result = MU_FAILURE;
+    }
+#ifdef SET_TRUSTED_CERT_IN_SAMPLES
+    // Setting the Trusted Certificate.  This is only necessary on system with without built in certificate stores.
+    else if ((iothubResult = IoTHubDeviceClient_SetOption(deviceHandle, OPTION_TRUSTED_CERT, certificates)) != IOTHUB_CLIENT_OK)
+    {
+        LogError("Unable to set auto Url encode option, error=%d", iothubResult);
+        result = MU_FAILURE;
+    }
+#endif // SET_TRUSTED_CERT_IN_SAMPLES
+    else
+    {
+        result = 0;
+    }
+
+    if ((result != 0) && (deviceHandle != NULL))
+    {
+        IoTHubDeviceClient_Destroy(deviceHandle);
+        deviceHandle = NULL;
+    }
+
+    if ((result != 0) &&  (iothubInitFailed == false))
```

Comment:
>if ((result != 0) &&  (iothubInitFailed == false)) [](start = 4, length = 50)

How about using:

if ((result != 0) && (iothubInitResult == 0))
{
IoTHub_Deinit();
}

then you do not use this additional iothubInitFailed

Reply:
Yeah I didn't like how that looked either, really like this way better.